### ISSUE TITLE
coqdep does not work on installed standard library

### ISSUE THREAD

**Issue by LasseBlaauwbroek** [coq/coq#11850]
#### Description of the problem

With a normal (non-local) Coq installation, running the following
```bash
coqdep $(COQPATH)theories/Init/Prelude.v
```
Results in these warnings:
```
*** Warning: in file theories/Init/Prelude.v, declared ML module cc_plugin has not been found!
*** Warning: in file theories/Init/Prelude.v, declared ML module ground_plugin has not been found!
*** Warning: in file theories/Init/Prelude.v, declared ML module numeral_notation_plugin has not been found!
*** Warning: in file theories/Init/Prelude.v, declared ML module string_notation_plugin has not been found!
```
This is because Coq's installer does not install `.mllib` files (but it does install `.v` files). The `coqdep` tool does not look for `.cmxs` files (or related) but rather the source `.mllib` files. This is rather inconvenient if I want to recompile a file in the standard library. May I suggest to either install the `.mllib` files or allow `coqdep` to look for `.cmxs` files?

#### Coq Version
The Coq Proof Assistant, version 8.11.0 (January 2020)
compiled on Jan 28 2020 22:33:30 with OCaml 4.09.0

**Comment by ejgallego** [comment]
This is almost a duplicate of https://github.com/coq/coq/issues/11026

Looking for cmxs files is not a solution as coqdep needs to work even before the ML part of files has been built. A proper solution for this is a bit complex, for now I'd recommend installing `.mlpack` files.

**Comment by LasseBlaauwbroek** [comment]
I don't immediately see a problem with `coqdep` first looking for a `.mllib` file and then as a fallback for corresponding binaries. This seems generally true for all files like `.v/.vo`, etc. After all, the dependency of a source file is not directly on another source file but rather on it's binary. I understand that it might be better to check the loadpaths for a source file first, but a fallback binary file search seems oke to me.

**Comment by ejgallego** [comment]
> the dependency of a source file is not directly on another source file but rather on it's binary

Source files don't depend on binaries, it is `.vo` files that depend on them. `.v` files depend on whathever is generating them, usually their own source file.

**Comment by LasseBlaauwbroek** [comment]
Right sorry, I meant 'binaries don't depend directly on a source file but rather on it's binary'.

**Comment by ejgallego** [comment]
> I don't immediately see a problem with `coqdep` first looking for a `.mllib` file and then as a fallback for corresponding binaries.

This seems problematic, in the sense that what should the build system do if there is a rule for the binary but it is still not built? The usual action here would be to build the binary, you don't want this to happen at the `coqdep` phase IMO.

**Comment by LasseBlaauwbroek** [comment]
It seems to me that this scenario cannot happen. If an applicable rule for the binary exists, then there must be a source `.mllib/.mlpack` file present to build the binary. Therefore, `coqdep` would not even be looking for the binary but simply select the source file (because looking for the binary is the fallback behavior).

**Comment by ejgallego** [comment]
> It seems to me that this scenario cannot happen. If an applicable rule for the binary exists, then there must be a source `.mllib/.mlpack` file present to build the binary. Therefore, `coqdep` would not even be looking for the binary but simply select the source file (because looking for the binary is the fallback behavior).

Not necessarily, for example the binary can come from an install rule, as indeed it does in your case.

**Comment by LasseBlaauwbroek** [comment]
Fair enough I guess. But I still don't really see a problem. I can make the same argument for source files. It can happen that `coqdep` is looking for a `.v` file that does not exist, but a rule to make it does exist (maybe this is a generated file, who knows). The current behavior of `coqdep` is not to go and build that file. So it would be completely consistent to also not build a binary during the `coqdep` phase.

**Comment by ejgallego** [comment]
The thing is that coqdep should be able to run very early, thus any solution that involves a binary being there seem incomplete or non-uniform; IMHO we should have the same rules in the build / installed case.

There are two things at discussion here:
- a hotfix, in this case we can either install the mlpack files or add a binary checking rule to coqdep
- the proper solution, which seems a bit more complex; likely proper scoping of plugins in the style proposed in #5028

**Comment by LasseBlaauwbroek** [comment]
Sure, and I don't really have a huge stake in this issue. For the hotfix, I would be happy to create a PR for either solution. Just let me know which one is preferred.

**Comment by ejgallego** [comment]
Ok, let me think for a bit what the right behavior should be.
